router/pkg/pubsub/nats/nats.go

```diff
@@ -22,35 +24,62 @@ var (
 )
 
 type connector struct {
-	conn   *nats.Conn
-	logger *zap.Logger
-	js     jetstream.JetStream
+	conn       *nats.Conn
+	logger     *zap.Logger
+	js         jetstream.JetStream
+	listenAddr string
 }
 
-func NewConnector(logger *zap.Logger, conn *nats.Conn, js jetstream.JetStream) pubsub_datasource.NatsConnector {
+func NewConnector(logger *zap.Logger, conn *nats.Conn, js jetstream.JetStream, listenAddr string) pubsub_datasource.NatsConnector {
 	return &connector{
-		conn:   conn,
-		logger: logger,
-		js:     js,
+		conn:       conn,
+		logger:     logger,
+		js:         js,
+		listenAddr: listenAddr,
 	}
 }
 
 func (c *connector) New(ctx context.Context) pubsub_datasource.NatsPubSub {
 	return &natsPubSub{
-		ctx:     ctx,
-		conn:    c.conn,
-		js:      c.js,
-		logger:  c.logger.With(zap.String("pubsub", "nats")),
-		closeWg: sync.WaitGroup{},
+		ctx:        ctx,
+		conn:       c.conn,
+		js:         c.js,
+		logger:     c.logger.With(zap.String("pubsub", "nats")),
+		closeWg:    sync.WaitGroup{},
+		listenAddr: c.listenAddr,
 	}
 }
 
 type natsPubSub struct {
-	ctx     context.Context
-	conn    *nats.Conn
-	logger  *zap.Logger
-	js      jetstream.JetStream
-	closeWg sync.WaitGroup
+	ctx        context.Context
+	conn       *nats.Conn
+	logger     *zap.Logger
+	js         jetstream.JetStream
+	closeWg    sync.WaitGroup
+	listenAddr string
+}
+
+// getInstanceIdentifier returns an identifier for the current instance.
+// We use the hostname and the address the router is listening on, which should provide a good representation
+// of what a unique instance is from the perspective of the client that has started a subscription to this instance
+// and want to restart the subscription after a failure on the client or router side.
+func (p *natsPubSub) getInstanceIdentifier() string {
+	hostName, _ := os.Hostname()
+
+	return fmt.Sprintf("%s-%s", hostName, p.listenAddr)
+}
+
+// getDurableConsumerName returns the durable consumer name based on the given subjects and the instance id
+// we need to make sure that the durable consumer name is unique for each instance and subjects to prevent
+// multiple routers from changing the same consumer, which would lead to message loss and wrong messages delivered
+// to the subscribers
+func (p *natsPubSub) getDurableConsumerName(durableName string, subjects []string) string {
+	subjHash := xxhash.New()
+	subjHash.WriteString(p.getInstanceIdentifier())
+	for _, subject := range subjects {
+		subjHash.WriteString(subject)
+	}
+	return fmt.Sprintf("%s-%x", durableName, subjHash.Sum64())
```

Comment:
`%x` is base16, are you sure it's not `%d` (base 10)?

Reply:
I made it like that on purpose, to have a shorter string

Reply:
Can that have an influence on collision?

Reply:
Shouldn't, because I'm just changing how I write the final hash, and not how it is calculated.